tensorly/backend/core.py

```diff
@@ -0,0 +1,854 @@
+import inspect
+import importlib
+import os
+import sys
+import threading
+import types
+import warnings
+
+import numpy as np
+import scipy.linalg
+import scipy.sparse.linalg
+
+
+_STATE = threading.local()
```

Comment:
> Why did you choose to specify the state as a thread-local data?

This is a thread-local because the sparse mechanism works by swapping the backends frequently, and global state should be threadsafe if possible. This means that tensorly can work with e.g. [dask](http://dask.pydata.org/) (my main project) easier, as it's already threadsafe.

>  Would there be any situation where we would want it/not want it to be local?

See comment on `set_default` below.

> Not sure whether we'd use multithreading but I guess in that case we'd have to create one of these per thread?

I'm not sure what you mean by this. All threads share a `threading.local()` object, but attributes dispatch to c-level threadlocals. If you're asking about tensorly backends specifically, the default backend propagates to newly created threads (there's a test for this too).

Reply:
> All threads share a threading.local() object

AFAIK the threading.local() object is **not** shared between all threads, the whole point being that the data is thread specific. 

Why not just store all this in a global BackendManager rather than threading.local()?
Though if this is crucial to your applications, we could keep it like this.. 

> the sparse mechanism works by swapping the backends frequently

While an easy way to handle the sparse case, in the end, this might be more of an issue than a feature..

Reply:
Coming back to this code after a few weeks, I'm thinking it might be best to do the following:

- `set_backend` is not a contextmanager, same api and rules as before
- backend dispatch internally supports a threadsafe context sensitive version (not publically exposed) that is used to make sparse functionality work.

> AFAIK the threading.local() object is not shared between all threads, the whole point being that the data is thread specific.

Apologies, I think we were talking past each other. Semantically yes they are different objects, implementation wise they are the same object which dispatches on attribute access to c threadlocal storage. Quibbles, not important.